### PR TITLE
In Progress: Remove post limit on sitemaps

### PR DESCRIPTION
**Pull request by nbloomf**
Addresses #3314: Remove 1000 post limit on sitemaps.
#### Changes proposed in this Pull Request:

Replace sitemap generating code.
#### Testing instructions:
1. Create a few test posts and upload some images.
2. Activate the sitemaps module of Jetpack.
3. Go to "example.com/news-sitemap.xml".
4. Go to "example.com/sitemap.xml". (It may take a refresh or two the first time.)

Note that sitemap.xml links to two other files; one is the root of a tree of sitemaps for pages, the other is the root of a tree of sitemaps for images. 

These changes allow for arbitrarily large sitemap trees per the [spec](sitemaps.org). To test this functionality without having to create an arbitrarily large site, adjust the constant SITEMAP_MAX_ITEMS in modules/sitemaps.php, (5 is a good value.) This constant determines the maximum number of URLs on a single sitemap, so that if it's set to 5 and we have 6 posts then the tree structure kicks in.

The constant SITEMAP_INTERVAL is the minimum number of seconds between rebuilds; it's currently set to 60 for testing.

We reuse the filter ``jetpack_sitemap_post_type`` to allow plugin authors to include custom post types in the sitemap.

#### Proposed changelog entry for your changes:
---
- [no] Make sure your changes respect [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- [NA] Did you make changes, or create a **new .js file**? If [Gulp](http://gulpjs.com/) is installed on your testing environment, run `gulp js:hint` before to commit your changes. It will allow you to [detect errors in Javascript files](http://jshint.com/about/).
- [NA] Did you create a **new action or filter**? [Add inline documentation](https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/php/#4-hooks-actions-and-filters) to help others understand how to use the action or the filter.
- [no] Create **[unit tests](https://github.com/Automattic/jetpack/tree/master/tests)** if you can. If you're not familiar with Unit Testing, you can check [this tutorial](https://pippinsplugins.com/series/unit-tests-wordpress-plugins/).

### CODE REVIEW

**Comment by nbloomf** [comment]
Closed for editing.

File: modules/sitemaps/sitemaps.php [hunk 1]
```diff
@@ -1,782 +1,2370 @@
...
- * @module sitemaps
- *
- * @link http://sitemaps.org/protocol.php Sitemaps.org protocol.
- */
-function jetpack_print_news_sitemap() {
 
-	$xml = get_transient( 'jetpack_news_sitemap' );
 
-	if ( $xml ) {
-		header( 'Content-Type: applicati
```
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Maybe this can be loaded (and any related code executed) only when `WP_DEBUG` is defined and set to `true`?

**Comment by nbloomf** [comment]
Thanks so much @eliorivero for your comments! I'm starting to read through them now.

To help me diagnose the reason for the 404 I have a couple of questions.
1. Can you tell me if you were able to get the news-sitemap.xml file to load properly?
2. Was the 404 a WordPress 404 page, or was it an empty XML document with a 404 return code?

File: modules/sitemaps/sitemaps.php [hunk 2]
```diff
@@ -1,782 +1,2370 @@
...
+	private function image_sitemap_to_index_item ( $post ) {
+		$url = site_url() . '/' . $post->post_title . '.xml';
+
+		/*
+		 * Must use W3C Datetime format per the spec.
+		 * https://www.w3.org/TR/NOTE-datetime
+		 * Also recall that we stored the most recent modification time
+		 * among all the posts in this sitemap in post_date.
+		 */
+		$last_modified = str_replace( ' ', 'T', $post->post_date) . 'Z';
```
Comment (nbloomf) [inline modules/sitemaps/sitemaps.php]: Do you mean `$last_modified`? It is used in line 1194 when we interpolate `$xml`.

**Comment by eliorivero** [comment]
1. news-sitemap.xml did not load either
2. a WordPress 404 page
3. my local runs PHP 5.6
As a side note, I tested it with both pretty and default permalinks.

File: modules/sitemaps/sitemaps.php (hunk 2 above)
Comment (eliorivero) [inline modules/sitemaps/sitemaps.php]: Sorry, I meant [the one in line 1226](https://github.com/Automattic/jetpack/pull/5394/files/5c738d899f4b66ea2a57377eafc97beed24ea6d5#diff-22179fe52fb0ae0ad3a4dca1dda03586R1226). It's not used in the scope of the method `image_sitemap_to_index_item`.

**Comment by nbloomf** [comment]
Hmm... the WordPress 404 means that the URL intercept callback isn't working. My dev environment has PHP 7.0.11; tomorrow I will try using 5.6 and lower to see if I can reproduce the error.

The permalink settings shouldn't have an effect - the module hijacks the usual routing process on sitemap URLs.

**Comment by nbloomf** [comment]
@eliorivero : I think I found the reason why the module wasn't working for you. When you get a chance, see if it's working in the latest commit.

BTW, I think I messed up the commit history by merging Automattic/jetpack:master back into this branch and then reverting the merge.

**Comment by nbloomf** [comment]
I have messed something up while trying to merge commits back from Automattic/jetpack:master and now this repo is broken. I will open a new PR with a fresh fork. :(

**Pull request closed without merging**